File: crates/idakit-sys-codegen/src/domains/export.rs
use super::super::model::*;

/// The export (entry-point) domain: per-export scalar accessors plus the name and forwarder
/// strings, indexed `[0, export_qty)`. `export_qty` is a templated passthrough; the lookups are
/// hand-written in `facade/export_custom.cc` (a forwarder-less export legitimately `Err`s).
pub const EXPORT: Domain = Domain {
    name: "export",
    sdk_includes: &["<entry.hpp>", "<stdexcept>"],
    externs: &[],
    structs: &[],
    consts: &[],
    custom_tu: Some("facade/export_custom.cc"),
    body_helpers: None,
    fns: fns! {
        "Number of exported entry points in the database (`get_entry_qty`)."
            export_qty() -> Usize = scalar("get_entry_qty()");
        "Address of export `idx`, or `BADADDR` when the export is a pure forwarder."
            export_ea(idx: Usize) -> U64;
        "Ordinal of export `idx`."
            export_ordinal(idx: Usize) -> U64;
        "Name of export `idx`; `Err` when it has none."
            export_name(idx: Usize) -> ResultString;
        "Forwarder target of export `idx`; `Err` when it has none (most exports do not)."
            export_forwarder(idx: Usize) -> ResultString;
    },
};

File: crates/idakit-sys/facade/export.cpp
// Hand-written Custom bodies for the generated export domain (namespace gen): entry-point
// lookups over get_entry/get_entry_ordinal, plus the name and forwarder as a rust::String
// (throws when absent, since most exports have no forwarder). export_qty is templated
// (gen_export_bodies.cc), not here.

#include <ida.hpp>
#include <pro.h>

#include <entry.hpp>

#include <stdexcept>

#include "gen_export.h"

namespace gen {

// Address of the entry point at idx in the entry-point list.
uint64_t export_ea(size_t idx) { return static_cast<uint64_t>(get_entry(get_entry_ordinal(idx))); }

// Ordinal of the entry point at idx in the entry-point list.
uint64_t export_ordinal(size_t idx) { return static_cast<uint64_t>(get_entry_ordinal(idx)); }

// Name of the export at idx; throws when the entry has no name.
rust::String export_name(size_t idx) {
  qstring out;
  if (get_entry_name(&out, get_entry_ordinal(idx)) <= 0)
    throw std::runtime_error("no export name at index");
  return to_rust_string(out);
}

// Forwarder target of the export at idx; throws when it doesn't forward (most exports don't).
rust::String export_forwarder(size_t idx) {
  qstring out;
  if (get_entry_forwarder(&out, get_entry_ordinal(idx)) <= 0)
    throw std::runtime_error("no export forwarder at index");
  return to_rust_string(out);
}

} // namespace gen
